services/paymaster/src/constraint.rs

```diff
@@ -526,8 +543,67 @@ impl DataConstraint {
     }
 }
 
+fn recover_signer_pubkey(signed_quote: SignedQuote) -> Result<H160, (StatusCode, String)> {
+    use anchor_lang::AnchorSerialize;
+
+    let signature_full = signed_quote.signature;
+    let signed_quote_bytes = signed_quote.try_to_vec().map_err(|e| {
+        (
+            StatusCode::BAD_REQUEST,
+            format!("Failed to serialize signed quote for message hashing: {e}"),
+        )
+    })?;
+    let message = signed_quote_bytes.get(..100).ok_or_else(|| {
+        (
+            StatusCode::BAD_REQUEST,
+            "Signed quote bytes too short to extract message for hashing".into(),
+        )
+    })?;
+
+    let msg = libsecp256k1::Message::parse(&keccak::hash(message).0);
+
+    let (signature, recovery_index_unnormalized) = signature_full.split_at(64);
+    let sig = libsecp256k1::Signature::parse_standard_slice(signature)
+        .map_err(|e| (StatusCode::BAD_REQUEST, format!("Invalid signature: {e}")))?;
+
+    let recovery_index_unnormalized_value = *recovery_index_unnormalized
+        .first()
+        .expect("Recovery index byte should be in the signature");
+    let recovery_id = libsecp256k1::RecoveryId::parse_rpc(recovery_index_unnormalized_value)
+        .map_err(|e| {
+            (
+                StatusCode::BAD_REQUEST,
+                format!("Invalid recovery index in signature: {e}"),
+            )
+        })?;
+
+    let secp_pubkey = libsecp256k1::recover(&msg, &sig, &recovery_id).map_err(|e| {
+        (
+            StatusCode::BAD_REQUEST,
+            format!("Failed to recover secp256k1 public key: {e}"),
+        )
+    })?;
+
+    let pubkey_hashed = keccak::hash(&secp_pubkey.serialize());
```

Comment:
do you need `&secp_pubkey.serialize()[1..65]` based on https://docs.rs/solana-secp256k1-recover/latest/src/solana_secp256k1_recover/lib.rs.html#446 ?

Reply:
goated catch!